core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -261,14 +283,38 @@ final class ShuffleBlockFetcherIterator(
             results.put(new SuccessFetchResult(BlockId(blockId), infoMap(blockId)._2,
               address, infoMap(blockId)._1, buf, remainingBlocks.isEmpty))
             logDebug("remainingBlocks: " + remainingBlocks)
+            enqueueDeferredFetchRequestIfNecessary()
           }
         }
         logTrace(s"Got remote block $blockId after ${Utils.getUsedTimeNs(startTimeNs)}")
       }
 
       override def onBlockFetchFailure(blockId: String, e: Throwable): Unit = {
         logError(s"Failed to get block(s) from ${req.address.host}:${req.address.port}", e)
-        results.put(new FailureFetchResult(BlockId(blockId), infoMap(blockId)._2, address, e))
+        val (size, mapIndex) = infoMap(blockId)
+        e match {
+          // Catching OOM and do something based on it is only a workaround for handling the
+          // Netty OOM issue, which is not the best way towards memory management. We can
+          // get rid of it when we find a way to manage Netty's memory precisely.
```

Comment:
Can we describe the entire process of handling OOM in the comment here? e.g. how we delay other fetch requests, etc.